### ISSUE TITLE
Include generated object files in commit history and published packages

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-go-instrumentation#1233]
This issue has been discussed before, I can't seem to find where though. The goal would be to have the released `go.opentelemetry.io/auto` package be a self-contained distribution.

Currently, when a project takes a dependency on `go.opentelemetry.io/auto` the project will not compile:

```go
package main

import "go.opentelemetry.io/auto"

func main() {
	i := auto.NewInstrumentation()
	_ = i.Close()
}
```

```terminal
$ go version
go version go1.23.2 linux/amd64

$ cat go.mod
module testing/autoimprt

go 1.23.2

require go.opentelemetry.io/auto v0.16.0-alpha

require (
	github.com/beorn7/perks v1.0.1 // indirect
	github.com/cenkalti/backoff/v4 v4.3.0 // indirect
	github.com/cespare/xxhash/v2 v2.3.0 // indirect
	github.com/cilium/ebpf v0.16.0 // indirect
	github.com/go-logr/logr v1.4.2 // indirect
	github.com/go-logr/stdr v1.2.2 // indirect
	github.com/google/uuid v1.6.0 // indirect
	github.com/grpc-ecosystem/grpc-gateway/v2 v2.22.0 // indirect
	github.com/hashicorp/go-version v1.7.0 // indirect
	github.com/klauspost/compress v1.17.11 // indirect
	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822 // indirect
	github.com/pkg/errors v0.9.1 // indirect
	github.com/prometheus/client_golang v1.20.4 // indirect
	github.com/prometheus/client_model v0.6.1 // indirect
	github.com/prometheus/common v0.60.0 // indirect
	github.com/prometheus/procfs v0.15.1 // indirect
	go.opentelemetry.io/auto/sdk v0.1.0-alpha // indirect
	go.opentelemetry.io/contrib/bridges/prometheus v0.56.0 // indirect
	go.opentelemetry.io/contrib/exporters/autoexport v0.56.0 // indirect
	go.opentelemetry.io/otel v1.31.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc v0.7.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp v0.7.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc v1.31.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp v1.31.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.31.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc v1.31.0 // indirect
	go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp v1.31.0 // indirect
	go.opentelemetry.io/otel/exporters/prometheus v0.53.0 // indirect
	go.opentelemetry.io/otel/exporters/stdout/stdoutlog v0.7.0 // indirect
	go.opentelemetry.io/otel/exporters/stdout/stdoutmetric v1.31.0 // indirect
	go.opentelemetry.io/otel/exporters/stdout/stdouttrace v1.31.0 // indirect
	go.opentelemetry.io/otel/log v0.7.0 // indirect
	go.opentelemetry.io/otel/metric v1.31.0 // indirect
	go.opentelemetry.io/otel/sdk v1.31.0 // indirect
	go.opentelemetry.io/otel/sdk/log v0.7.0 // indirect
	go.opentelemetry.io/otel/sdk/metric v1.31.0 // indirect
	go.opentelemetry.io/otel/trace v1.31.0 // indirect
	go.opentelemetry.io/proto/otlp v1.3.1 // indirect
	golang.org/x/arch v0.11.0 // indirect
	golang.org/x/exp v0.0.0-20230224173230-c95f2b4c22f2 // indirect
	golang.org/x/net v0.30.0 // indirect
	golang.org/x/sys v0.26.0 // indirect
	golang.org/x/text v0.19.0 // indirect
	google.golang.org/genproto/googleapis/api v0.0.0-20241007155032-5fefd90f89a9 // indirect
	google.golang.org/genproto/googleapis/rpc v0.0.0-20241007155032-5fefd90f89a9 // indirect
	google.golang.org/grpc v1.67.1 // indirect
	google.golang.org/protobuf v1.35.1 // indirect
)

$ go build
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/database/sql/bpf_x86_bpfel.go:165:12: pattern bpf_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/github.com/segmentio/kafka-go/consumer/bpf_x86_bpfel.go:172:12: pattern bpf_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/github.com/segmentio/kafka-go/producer/bpf_x86_bpfel.go:167:12: pattern bpf_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/go.opentelemetry.io/auto/sdk/bpf_x86_bpfel.go:159:12: pattern bpf_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/go.opentelemetry.io/otel/traceglobal/bpf_x86_bpfel.go:213:12: pattern bpf_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/google.golang.org/grpc/client/bpf_x86_bpfel.go:170:12: pattern bpf_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/bpf_x86_bpfel.go:169:12: pattern bpf_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/net/http/client/bpf_no_tp_x86_bpfel.go:182:12: pattern bpf_no_tp_x86_bpfel.o: no matching files found
../../../pkg/mod/go.opentelemetry.io/auto@v0.16.0-alpha/internal/pkg/instrumentation/bpf/net/http/server/bpf_x86_bpfel.go:174:12: pattern bpf_x86_bpfel.o: no matching files found
```

We should ship a Go module/package that is self-contained. It should not require any user to generate code within their package cache.

cc @damemi @RonFed

**Comment by damemi** [comment]
https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/695

I’m not really in favor of this without a really strong reason due to the
complications that binary files create with Git. I don’t think that is
worth the convenience over (for example) a generator that can be run via
Make, so I see it more as a tooling issue than a library issue.

**Comment by MrAlias** [comment]
The thing is, you cannot execute a make task when someone downloads the package. What we ship now is a broken package.

**Comment by damemi** [comment]
I meant a make task to run go generate in the module cache. Which is not
pretty, but I would like us to look into other ways to make this easier
before we commit to committing the object files directly.

The BPF object files themselves aren’t code, they’re really release
artifacts. The code manages and depends on these artifacts to run, but to
me they seem kind of like DLLs in that sense. So, the package isn’t broken,
it’s just missing dependencies by default.

It looks like others have had this question before (
https://github.com/cilium/ebpf/discussions/988 with an interesting
solution). I don’t disagree that we should do *something* to make this
easier to use, but I’m hesitant to commit them.

**Comment by MrAlias** [comment]
> [cilium/ebpf#988](https://github.com/cilium/ebpf/discussions/988)

This is talking about running a make target within the project generating the eBPF code. It is not talking about providing a release target to others downloading the Go module. That approach is going to be flawed as it will require someone who runs `go get`  or `go mod tidy` in their project to then go find our distributed `Makefile` and run another command. This is going to be especially problematic if the `auto` package is not even a direct dependency.

**Comment by MrAlias** [comment]
The code that we ship already directly depends on the files we do not release:

https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/77e8de9e0f4e33afb00bacb77917bad4f3056d1f/internal/pkg/instrumentation/bpf/database/sql/bpf_x86_bpfel.go#L163-L166

This seems like something that should be in source code if we are already referencing it.

**Comment by damemi** [comment]
I’m saying we should release those files, but I don’t think publishing them
in a Git commit is the best way to do that.

Are there other (non eBPF) Go packages that embed binary files and commit
them? I’m wondering if there is any precedent for this pattern outside of
eBPF or if we would just be trading one unusual situation for another.

Within eBPF-Go there isn’t much guidance. I linked that other issue to show
that the main problem (is it idiomatic to commit BPF objects to git) is
still pretty unanswered. Maybe this is a gap in the Cilium/bpf2go projects
themselves.

Again, I’m totally with you on publishing these files but I just want to
pause and see if we can think of a better approach, if possible.

**Comment by MrAlias** [comment]
Looking at the [remove import paths documentation](https://pkg.go.dev/cmd/go#hdr-Remote_import_paths), we need to use some VCS as the protocol to download source code (i.e. commit all the code), or run our own [module proxy](https://go.dev/ref/mod#goproxy-protocol). Looking at the documentation for module proxies, it seems like we could respond to `$base/$module/@v/$version.zip` requests with a zip file including all the generated code.

This module proxy will need to be added as a managed service run by OTel and require additional step in our release process for uploading the release zip files.

**Comment by MrAlias** [comment]
FWIW, here's an example of another eBPF project checking in their object files: https://github.com/k8spacket/k8spacket/blob/65e5268975cc41a4a1b5daed8a534666a8a4fab5/ebpf/tc/tc_bpfel.o

**Comment by RonFed** [comment]
I'm very supportive of improving this part - I agree with @damemi - committing these files to git should only be done if we can't find another reasonable approach.
In Odigos, we [call make generate](https://github.com/odigos-io/odigos/blob/509afdc1a010e91cc22358a2f862503b5b9c495d/odiglet/Makefile#L9-L14).
From my experience, besides this being not pretty it also makes development harder since each import of our library results in a build error - which makes testing harder.

**Comment by MrAlias** [comment]
Another idea: we could make our releases in release branches. Those branches would contained checked-in `*.o` files. If I understand the GitHub Go releasing model correctly (a big if), we might be able to delete the branch after the release is tagged and published.

I'm not sure if this is better or just the same with regard to binary data being checked in to git.

**Comment by MrAlias** [comment]
We could also post-curate the `generate` steps with a file manipulator that directly defines the object file binary data in the embedded type. This would be equivalent to what protobuf does. For example: https://github.com/open-telemetry/opentelemetry-proto-go/blob/53975ec09b7a48899d5153f09425f326e642a2dc/otlp/trace/v1/trace.pb.go#L963-L1119

**Comment by MrAlias** [comment]
@damemi @RonFed: please update the description with any additional proposals or pros/cons you can think of.

**Comment by damemi** [comment]
@MrAlias the post-curation idea is really interesting. That seems like the safest compromise and the pattern already exists for protobuf. I think that's my favorite option, and we could even separate the generated files into their own sub packages if we wanted.

Just thinking, you could make a philosophical argument here that while this is broken in the traditional Go sense, it's also not a very traditional Go library. I would absolutely expect almost any other Go package to not behave this way. But anyone who's using this, even transitively, should be aware that it depends on compiled eBPF objects... and so I wonder if the extra build/generate steps (like the make command @RonFed shared) could be an acceptable pattern? At least until some better guidance or alternative emerges?

To formalize that option, I'd propose our docs include a section on "using this library" that clearly includes exactly what commands the user needs as a setup step after they run `go get` (ideally a copy/paste one liner).

**Comment by RonFed** [comment]
> Post-process Go code to contain the object file

This object file representation will only be updated on releases or every time we change the c files?
I did not fully understand the advantages relative to just using the object files - is this not the same thing written differently? (i.e will have the same effect on git?)

> Run and manage our own Go module proxy

With this approach, if a user `go get` our module, the proxy server will provide the `.o` files as if there were in git?
If so, that sounds like a complete solution to me (which will probably require a lot of work though). 


A common question to most of the propsed soultions is whether we want to support imports of non-official releases (with commit hashes)? I think some of the options rely on just updating the compiled code in each official release.

**Comment by MrAlias** [comment]
> I did not fully understand the advantages relative to just using the object files - is this not the same thing written differently? (i.e will have the same effect on git?)

I don't thing there objectively is a difference. The difference is stylistic. There is precedence with protobuf generating binary data directly held in `.go` files that is checked into repositories.

If others are okay with this approach, I'm okay as well. It is not my first choice though.

**Comment by MrAlias** [comment]
> With this approach, if a user `go get` our module, the proxy server will provide the `.o` files as if there were in git?
> If so, that sounds like a complete solution to me (which will probably require a lot of work though).

Correct. I haven't built out a prototype yet, but all the documentation I have read indicates this will be the behavior. We will upload snapshots of our code (maybe already zipped(?)) to some module proxy server. Those snapshots will be served and they will contain all the `*.o` files. The git history would remain as it currently is in this approach.

**Comment by MrAlias** [comment]
> A common question to most of the propsed soultions is whether we want to support imports of non-official releases (with commit hashes)? I think some of the options rely on just updating the compiled code in each official release.

That's a good question. We would have to investigate this further to see what kind of support we could provide.

I know we can provide the standard Go proxy as a fallback to our proxy and it may all users to download directly from git history. The issue would be that those downloads would not have the desired object files.

I expect there is some more engineering we could do to try and solve the issue though.

**Comment by MrAlias** [comment]
Go mod proxies we could run BTW:

- https://github.com/gomods/athens
- https://pkg.go.dev/github.com/goproxy/goproxy
- JFrog -- https://gavincampbell.dev/post/golang-proxy-artifactory/

**Comment by damemi** [comment]
The problem I see with the Go proxy solution is it doesn't support custom Probes, and makes it tougher for us to do a contrib-style repo. I guess third party probes can choose to provide their object files however they'd like, but a consistent approach would be nice

**Comment by damemi** [comment]
Another suggestion from [ebpf slack](https://cilium.slack.com/archives/CKC55JL8L/p1730498226979079?thread_ts=1730324684.565339&cid=CKC55JL8L) was to use Git LFS, which from a quick search looks like it should be compatible with Go modules. I don't know if that would help with the code reviews and diffs, but it would solve the repo bloat.

I also like the release branch idea, and I think we should start using release branches anyway for backports once our release model is more stable.

**Comment by MrAlias** [comment]
It looks like using Git LFS will mean all modules that take this one as a dependency will need to have it installed and configured in their development environments: https://stackoverflow.com/questions/58280942/import-a-golang-module-with-lfs-objects

**Comment by damemi** [comment]
> It looks like using Git LFS will mean all modules that take this one as a dependency will need to have it installed and configured in their development environments: https://stackoverflow.com/questions/58280942/import-a-golang-module-with-lfs-objects

Yeah, that is the downside. But I think it's still an improvement over the weird go cache generate commands you need otherwise since "all" you have to do is install git lfs.

By the way, when we finish the custom probes API https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1105 this should contain the blast radius of this issue to just the probe modules themselves, since at that point the main auto-instrumentation framework won't have an explicit dependency on the probes. The problem still exists, but I think that in conjunction with one of the other solutions is going to be enough

**Comment by damemi** [comment]
I also had an idea for a proposal we could make to the Cilium libraries to support runtime object storage:

The compiled object bytes aren’t actually a build dependency, they’re only needed at runtime when the Cilium libraries load them into the kernel. The file just needs to be there for the Go compiler to build.

So if we built a similar approach to Git LFS and had bpf2go embed a file that just held a pointer to the actual object, then there wouldn’t be any compile errors and you could support any kind of runtime object storage: local, remote, lfs, etc… just have Cilium read the bytes into memory at runtime. The pointer file format would just have to declare the protocol and location and use that if an actual local file isn’t available.

There’s details to work out in the UX, like for example when developing you would want to be able to build a local file that takes precedence over whatever the pointer file says.

The downside is that you would need a connection to wherever the remote file is. We could solve this by supporting a persistent local cache (ie, download it to /tmp and read from there) or even a runtime flag that points to a local directory where the objects live (for example, each release we just publish the objects as a GitHub artifact, users download the ones they want for the probes they’re using, and save that in their host/image/etc). The tradeoff there is file I/O, but you could cache it in memory for efficiency if it's going to be loading/unloading often.


Wdyt? I'm pretty sure this is feasible but I have to look more into the details

**Comment by MrAlias** [comment]
SIG notes:

- Each `*.o` file is ~50KB. Every change would mean another 50KB to the git history.

**Comment by damemi** [comment]
More updates from the slack thread:

* We should make sure we're using `llvm-strip` to reduce the size of our .o files
* Cilium already supports direct loading instead of embedding by calling `ebpf.LoadCollectionSpec` on the file and `spec.LoadAndAssign` on the go structs from bpf2go. If so, we could make this change today and immediately fix the IDE/compile time issue

**Comment by grcevski** [comment]
I just wanted to follow-up here and summarise all the things we've tried with Beyla and how it might integrate with the OpenTelemetry collector.

1. We currently use git-lfs and we store the binaries in the git repo. This has proven to be problematic for two reasons. One is that the project importing the eBPF enabled project must also use git-lfs and there seems to be a problem with the `go get` command and git-lft. We couldn't make it work, although it might be something we missed.
2. We tried embedding the binaries as bytes directly in the .go code generated by bpf2go, but this isn't a great solution since the files become even larger and often small changes in one place cause large diffs. So it didn't solve anything.
3. We are pursuing two new approaches:
    - Tooling that can be used by downstream projects to generate the binaries on demand after the `go get` command has run. This has the downside that it requires a change in the downstream project, but with the use of Docker we can make this nicely separated. It does impose a Docker build dependency on the downstream project. The alternative would be a clang/llvm dependency on the downstream project.
    - Don't commit the eBPF binaries on the main branch, but commit them on release branches. The reasoning here is that most downstream projects will import the released versions and the velocity of change to the binaries on release branches will be small. The commit of the binaries on the released branches can be done via CI job that will ensure the binaries match the source eBPF programs and haven't been tampered with. At the same time, the main branch which will see a lot of changes will stay clean and not grow unreasonably over time.

**Comment by damemi** [comment]
@grcevski those are my favorite 2 options we've come across so far too. I think either one is justified for a project like this. The tooling option could also be based on go workspaces, if you haven't looked into that already. That's how I work on Odigos, I just have this repo checked out, generate the files locally, and point to it in my `go.work`

**Comment by grcevski** [comment]
> The tooling option could also be based on go workspaces

Oh that's cool, we haven't looked at that!

**Comment by damemi** [comment]
The tricky part with go workspaces is it's really meant for local development (meaning it can be different for every developer based on their environment and folder structure), so it's not recommended to check it into git and it makes generating a "default" tricky

**Comment by MrAlias** [comment]
SIG meeting notes:

- Ask the go tooling team if they have options or are willing to look into solutions.

**Comment by RonFed** [comment]
As mentioned in https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2449183152, in Odigos we are doing what Rafeal described above in the second point.

The `.o` files don't have to be in the same location as the source code. We could use [LoadCollectionSpecFromReader](https://pkg.go.dev/github.com/cilium/ebpf@v0.17.3#LoadCollectionSpecFromReader) directly without relying on `bpf2go` (don't really have a solution, just bringing this up as it might help getting to one)

**Comment by rafaelroquetto** [comment]
@RonFed That is indeed a good point (not relying on `bpf2go`) - I had done something similar with Beyla but dropped it because there was so much boilerplate being manually written (and therefore very error prone). Perhaps we could look into extending `bpf2go` itself to add a new flag telling it where to place the .go files and where to place the .o files - although in this case, we'd still be committing the generated `.go` files - we gotta make sure they don't drift from the `.o` files... maybe something to think about, I don't know...

**Comment by rafaelroquetto** [comment]
To expand a bit on https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2698298287, I wanted to elaborate a bit on the machinery behind `beyla_genfiles.go`. Since that comment, there were a few improvements, and I've toyed with it again using Grafana Alloy as a playground (as it's vendoring Beyla) IMHO presenting satisfactory results.

I think it's easier to illustrate it top-down from the [corresponding Alloy PR](https://github.com/grafana/alloy/pull/2734). The canonical way to vendor Beyla is to simply add a "generate" step before the compilation step. In the case of Alloy, that meant a new Makefile target called `generate-beyla` which invokes the artifact generation. [The relevant target is here](https://github.com/grafana/alloy/blob/aee82782a4b9093ca485eaa1b92ee9fdcdbe4959/Makefile#L306).

Everything else revolves around it.

Unlike building artifacts on your own project, building artifacts of a dependency means finding it, scanning what needs to be generated and triggering the generate process. This is why `beyla_genfiles.go` exists. For instance, you don't want to keep regenerating Beyla's artifacts everytime you, as an end-developer, is building Alloy - so `beyla_genfiles` takes care of that _for ebpf files only_ (we delegate to the generate command itself for other filetypes to decide whether to rebuild - [here's one such example using wget -N flag](https://github.com/grafana/beyla/blob/6be5ee6cde26bf6535d5c0498b6bdfa95c0ad0cf/pkg/internal/otelsdk/sdk_inject.go#L5) It also takes care of invoking `bpf2go` in parallel (or rather, `go generate`), to speed things up.

The idea is to provide the end-user/developer with the least steps possible, so the Makefile target simply invoke beyla_genfiles as a tool: 
```
generate-beyla:
  go run github.com/grafana/beyla/v2/cmd/beyla-genfiles@vX.Y.Z
```
If you look at [beyla_genfiles.go](https://github.com/grafana/beyla/blob/main/cmd/beyla-genfiles/beyla_genfiles.go), a lot of the code is dealing with managing GitHub workflow builds. `beyla_genfiles` has two modes of operation:

1. building things locally: this is the "simpler" way in which it simply gathers the files to generate and generates them (in the right place). This requires the user to have the appropriate tooling deployed (e.g. clang and bpf2go) and is what Beyla developers use day-to-day. 
2. building things using a container: in this mode, beyla_genfiles fetches a container image (configurable), mounts the Beyla module root inside the container with the correct user permissions, and then re-runs beyla_genfiles inside the container using the "local build mode".  This is what a downstream project like Alloy would use by default, so that we don't impose tooling requirement downstream. This is also what GH uses and ensures binaries are consistent. Because of that, `beyla_genfiles` needs to know how to properly setup the volume for the generator image from the checkout GH source tree, and this is why there's some GH logic inside beyla_genfiles for resolving paths (there are differences when GH containers are involved).

Finally, the last missing bit is the generator image that packs `beyla_genfiles` for the point number 2 above. The [Dockerfile](https://github.com/grafana/beyla/blob/main/generator.Dockerfile). `beyla_genfiles` will fetch that image, correctly resolve the local Beyla module path and mount it to the container's `/src/` volume, with the right permissions - the container then invokes `beyla_genfiles` itself internally in "local generation mode" and voi-la - the files are generated.

To sum up, for a user wanting to build Alloy, all they need to do is:
```
git clone https://github.com/grafana/alloy/
cd alloy && make alloy
```

And files will be automagically generated.

Happy to do a walk-through/quick demo on the next SIG meeting.

**Comment by MrAlias** [comment]
> If you look at [beyla_genfiles.go](https://github.com/grafana/beyla/blob/main/cmd/beyla-genfiles/beyla_genfiles.go), a lot of the code is dealing with managing GitHub workflow builds. `beyla_genfiles` has two modes of operation:

New gen files: https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/db2937ddd0378b8e2d0f17c3f8204c9a5c8d1451/cmd/obi-genfiles/obi_genfiles.go

**Comment by rafaelroquetto** [comment]
In addition to the `obi-genfiles` approach, in OBI we currently use a git-submodule approach, which can feel a bit alien but it turns out to be very flexible.

To recap the issue: the go tooling doesn't play well with dependencies that need to generate artefacts as part of their build process (e.g. generating ebpf object files, or any code generation really). The reason is that these generated artefacts need to be part of their corresponding go package, and as such, need to end up in the package dir itself, which, in the case of dependencies, means the go module / package directories - these are theoretically meant to be immutable, and `obi-genfiles` works around (and violates) that.

The git-submodule approach is a simple and more canonical approach: it relies on using git submodules for dependencies, a common practice outside of the golang world. Taking Grafana Beyla, which depends on OBI, to illustrate, all we need to do is:

1. add a git submodule with the OBI dependency
2. trigger the object file generation (or code generation) inside the submodule, i.e. as if we were building in-tree, the usual deal
3. replace the go.mod dependency to to point to that submodule path: `replace github.com/open-telemetry/opentelemetry-ebpf-instrumentation => ./.obi-src`
4. `go mod vendor` - at this stage, you end up with the vendored dependency containing all the generated code.

The usage of submodules has the additional benefits in which is very flexible - any submodule in any branch can point to any SHA1 - so it's very easy to pin your project to a given upstream version, and update it at your own pace. The learning curve is also very easy, as the workflow is a simply git submodule workflow.

**Comment by MrAlias** [comment]
**tl;dr**: Running a Go module proxy is technically possible, but the security and stability risks introduced make it an undesirable solution.

I have put together [a proof-of-concept](https://github.com/MrAlias/bpfgenproxy) for how we can run a [Go module proxy](https://go.dev/ref/mod#goproxy-protocol) along with our [vanity URL server](https://github.com/open-telemetry/opentelemetry-go-vanityurls/tree/main) to serve custom zip files of our project that include the generated eBPF object files without having them in our repository history.

## Overview

```diagram
+--------+                                              +-------------------+
|        |-- 1. go mod tidy --------------------------->|                   |
|        |                                              |                   |
|        |                                              | Go Proxy (Vanity) |
|        |                                              |                   |
|        |<- 2. "http://goproxy.opentelemetry.io/auto" -|                   |
|        |                                              +-------------------+
| Client |
|        |                                              +-------------------+
|        |-- 3. Client requests module from proxy ----->|                   |
|        |                                              |                   |
|        |                                              | Module Proxy      |
|        |                                              |                   |
|        |<- 4. Zip file of go.opentelemetry.io/auto ---|                   |
+--------+                                              +-------------------+
```

1. The Go client runs `go mod tidy` and requests the module information for `go.opentelemetry.io/auto` from the Go proxy server serving the vanity URL.
2. The Go proxy server responds, pointing the client to the module proxy and tells the client to use the `mod` protocol for the download (instead of a VCS).
3. The client requests the module from from the module proxy.
4. The module proxy serves a cached copy of the module, which includes pre-generated eBPF object files.

All other modules requested from the module proxy are treated as pass-through requests to the [upstream Go proxy](https://proxy.golang.org).

See [the repository](https://github.com/MrAlias/bpfgenproxy) for specifics on how this works and the code so you can run the example.

## Takeaways

### This will be an attack vector

We are dynamically generating content for Go modules that is not present in the upstream repository.
This is an attack vector for malicious actors to exploit.

We do not review the content of the modules served by the module proxy.
If a malicious actor were to upload/generate a module with malicious code, it could compromise the systems that use this proxy (i.e. all of our direct and indirect users).

### The `go.sum` file

The `go.sum` file is used to verify the integrity of the downloaded modules.
It contains checksums for each module version, ensuring that the content has not been tampered with.
This is crucial for security.

All modules served by the module proxy are generated dynamically.
They are guaranteed to be different from the upstream Go proxy given they contain the additional eBPF object files.

That means that the `go.sum` file will not match the checksums of the modules served by the upstream Go proxy if the VCS system was originally used.
Consequently, this means that we will only ever be able to serve modules that are not already present in the upstream Go proxy without causing security errors in the Go client.

Additionally, this means that we need ways to retain all generated zip files in the module proxy in a reliable way.
If they are not retained, and regenerated, there is significant chance the hash will change and cause security errors in the Go client.

**Comment by MrAlias** [comment]
I'm planning to evaluate ["Commit the object files into release branches"](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2450723817) next.

**Comment by mx-psi** [comment]
Copy-pasting some questions I have about https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-3119891483 with regards to the discussion over at open-telemetry/opentelemetry-collector/issues/15430.

> We are dynamically generating content for Go modules that is not present in the upstream repository.

One thing I am failing to understand is why you decided to go with a dynamic server when my understanding is that a Go module proxy can be served with a static website (see [first sentence here](https://go.dev/ref/mod#goproxy-protocol)). Is it a storage concern?

> This is an attack vector for malicious actors to exploit.

Also for my understanding: Github releases are _also_ not present on the repository: what are the additional security guarantees that are provided by using them instead of using a static service?

> They are guaranteed to be different from the upstream Go proxy given they contain the additional eBPF object files. [...] Consequently, this means that we will only ever be able to serve modules that are not already present in the upstream Go proxy without causing security errors in the Go client.

I do not understand why it is a requirement to serve older modules. Could you elaborate on that?

> Additionally, this means that we need ways to retain all generated zip files in the module proxy in a reliable way.

Right, this was my assumption on how the proxy would be implemented: as a static server that serves the generated zip files.

**Comment by MrAlias** [comment]
Thanks for moving the proxy-specific questions here. I agree this is the right place to discuss whether a Go module proxy is a viable solution for generated artifacts behind `go.opentelemetry.io/...` import paths.

I want to separate two questions, though:

1. Is a Go module proxy technically possible?
2. Should the possibility of a Go module proxy block open-telemetry/opentelemetry-collector#15430?

For (1), yes, it is technically possible. That was the point of https://github.com/MrAlias/bpfgenproxy. The PoC demonstrates that the vanity URL service can return `vcs=mod` for `go.opentelemetry.io/auto`, that the Go tool can then ask a module proxy for the module, and that the proxy can return a zip containing the generated object files.

However, the PoC also shows the parts that still need production design. It special-cases `go.opentelemetry.io/auto@v0.22.1`, serves local `info`, `mod`, and `zip` files for that single version, and falls back to the upstream Go fetcher for other modules. The example compose file also sets `GONOSUMDB` for `go.opentelemetry.io` and has a TODO about avoiding that checksum bypass. That is enough to prove the mechanics. It is not yet an operational proposal for OpenTelemetry module distribution.

It is also important context that this has not been a drive-by decision from one project. The maintainers and contributors working on OpenTelemetry eBPF projects have been discussing this class of problem for a long time across `opentelemetry-go-instrumentation`, OBI, Beyla experience, and related SIG conversations. The options discussed include checking in object files, release branches, post-processing generated Go, Git LFS, downstream generation tooling, submodules, source archives, and a Go module proxy. The current direction is the result of that project-maintainer tradeoff discussion, not a lack of awareness that a module proxy exists.

> One thing I am failing to understand is why you decided to go with a dynamic server when my understanding is that a Go module proxy can be served with a static website. Is it a storage concern?

I do not think "static website" avoids the dynamic-generation problem. A static Go module proxy can serve already-materialized protocol files, but those files still have to be generated first: `.info`, `.mod`, `.zip`, `/@v/list`, and `/@latest`.

That matters because Go modules are not only downloaded at release tags. The Go command can resolve pseudo-versions for individual commits. If we want the module path to work as a transparent replacement for VCS access, then every supported commit that a user can request needs a generated module representation containing the generated artifacts. That can happen eagerly on every upstream push, lazily on first request, or only for a restricted set of versions, but each choice has consequences.

If we generate eagerly, then we need a pipeline that produces, verifies, publishes, and permanently retains module proxy artifacts for every relevant commit. If we generate lazily, then the proxy is doing dynamic generation at request time or in response to request pressure, which is exactly the operational and security concern from the original takeaway. If we restrict the proxy to only preselected releases, then it is no longer a transparent Go module source; it is a narrower release-artifact distribution mechanism, and we still need to define what happens for pseudo-versions, unreleased commits, and already-published versions.

So the hard part is not whether the final bytes can be served from a static filesystem. Of course they can. The hard part is the artifact-production system behind that filesystem: what versions it covers, when it generates them, how it proves they match source, how it prevents tampering, how it publishes them immutably, and how it behaves when users request versions that have not been precomputed.

This brings back the security concern instead of removing it. A static host for generated module zips is still a high-value module-origin target if it is the thing users trust for `go.opentelemetry.io/...` module contents. The generation and publication pipeline becomes the place where malicious code or mismatched generated artifacts could be introduced, and the static host becomes the place where those artifacts are distributed to downstream users.

That is why I do not think it is appropriate for the Collector discussion to turn into "the eBPF projects should solve this by hosting a Go module proxy" unless someone is prepared to own that design and operational burden. Collector maintainers and approvers absolutely get to decide what belongs in OCB. But deciding that OBI, `go.opentelemetry.io/auto`, or other eBPF projects should publish through a new VCS/module-proxy path is a decision for those projects and the relevant OTel infrastructure owners, because they are the ones who would inherit the release, security, storage, and incident-response responsibilities.

> Github releases are also not present on the repository: what are the additional security guarantees that are provided by using them instead of using a static service?

I am not arguing that GitHub release assets are magically safer merely because they are GitHub release assets. The difference is the trust model.

The `source_archive` proposal in open-telemetry/opentelemetry-collector#15430 requires the OCB manifest to explicitly name an archive URL and pin its bytes with a SHA-256 digest. In the current draft implementation (open-telemetry/opentelemetry-collector#15431), OCB downloads that exact artifact, verifies the digest, extracts it into a local digest-keyed cache, reads the module path from the archive's `go.mod`, and wires a local `replace`. That is an explicit build input chosen by the collector distribution author.

There is also an existing GitHub-native verification path around the OBI release artifacts. The [OBI release process](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/main/RELEASING.md#automatic-release-workflow) runs the release checks, builds the `source-generated` archive from the tagged source plus generated artifacts, publishes SBOMs, publishes `SHA256SUMS`, and documents checksum verification for users. GitHub release asset metadata also records the SHA-256 digest of each uploaded release asset. For example, the [v0.9.0 release](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/releases/tag/v0.9.0) asset metadata includes a digest for `obi-v0.9.0-source-generated.tar.gz`, and GitHub's attestations API returns a release attestation for that digest whose subject set includes the source-generated archive, `SHA256SUMS`, SBOM assets, and binary release archives. That matters because consumers are not just trusting an arbitrary file server; they have digest-pinned artifacts, release metadata, GitHub release attestations, and SBOM/checksum assets attached to the published release.

GitHub's [artifact attestation model](https://docs.github.com/en/actions/concepts/security/artifact-attestations) is explicitly about creating signed provenance and integrity claims for release artifacts so consumers can verify where artifacts came from and make policy decisions. That does not prove an artifact is safe by itself, but it is a real verification pipeline. A new Go module proxy would not inherit that automatically. If the alternative is "serve generated module zips from a static Go proxy", then that proposal also needs to explain what the equivalent provenance, attestation, checksum, SBOM, and user-verification story is for every generated module zip it publishes.

A Go module proxy has a different trust model. It changes what bytes are served behind a Go module path/version. That can be fine, but it makes the proxy and its generation pipeline part of the module's publication infrastructure. If the proxy, generated artifact store, or publication pipeline is compromised, misconfigured, or loses historical artifacts, every direct or indirect user resolving through that module path is affected. The Go module checksum database exists specifically because untrusted proxies should not be able to serve different code unnoticed.

So the security comparison is not "GitHub release asset versus static file server". It is "explicit URL plus digest in a build manifest, backed by the existing OBI release artifact verification path" versus "new OTel module-origin infrastructure that must preserve Go module immutability, checksum semantics, generation provenance, artifact attestations, SBOM/checksum publication, and artifact retention".

> I do not understand why it is a requirement to serve older modules. Could you elaborate on that?

It is not a requirement for the `source_archive` proposal. It is a requirement that has to be addressed if the Go module proxy is proposed as the general replacement for that proposal.

Existing module versions may already be known to `proxy.golang.org`, `sum.golang.org`, and users' `go.sum` files with the contents fetched from VCS. If OTel later serves different zip bytes for the same module path/version, Go clients can reject the download because the checksum does not match the bytes already associated with that version.

A future-only proxy might be possible, but that still needs a concrete design: where is the cutoff, how do we prevent a future release or pseudo-version from being observed by the public proxy before the generated-artifact zip is published, what happens to existing broken versions, and what environment variables or checksum-db policy are required for users? The bpfgenproxy PoC does not settle those questions.

Also, "OCB versions should align with Collector Go API versions" does not answer this. The checksum issue is about module identity and immutable bytes for a module path/version, not support policy.

> Right, this was my assumption on how the proxy would be implemented: as a static server that serves the generated zip files.

That may be a viable serving layer, but it is not a complete solution. If you believe this is the better path, I think the next useful step is for you to propose it concretely: a design issue or PR for the relevant OTel infrastructure that covers the generation policy, serving layout, release process, pseudo-version behavior, checksum behavior, artifact attestations, SBOM/checksum publication, immutability/retention guarantees, ownership, and failure modes. Ideally it should include a working prototype for at least OBI or `go.opentelemetry.io/auto`, with default Go checksum behavior documented, or an explicit argument for any required `GONOSUMDB`/`GOSUMDB` policy.

Until that exists, I do not think open-telemetry/opentelemetry-collector#15430 should be blocked on the possibility of a Go module proxy, and I do not think the Collector issue should be used to override the distribution choices made by the eBPF project maintainers. The Collector issue is asking a narrower question: should OCB support an explicitly declared, checksum-pinned source archive as component source? That proposal can be reviewed on its own schema, security, and builder-maintenance merits.

If the static Go proxy proposal becomes concrete and maintainers prefer it, great; we can compare it against `source_archive`. But right now it is an unowned alternative with unresolved module-distribution, generation, and checksum questions. I do not think that should be enough to stop progress on a concrete OCB proposal.

**Comment by mx-psi** [comment]
Thanks for the detailed answer!

> The current direction is the result of that project-maintainer tradeoff discussion, not a lack of awareness that a module proxy exists.

Of course, I did not mean to imply that :) I do think it is important that the Collector SIG understands what we are trying to accomplish here and what are the requirements so we can work together on a solution that has minimal maintenance burden for both parties.

> > One thing I am failing to understand is why you decided to go with a dynamic server when my understanding is that a Go module proxy can be served with a static website. Is it a storage concern?
> 
> [...] 
> If we restrict the proxy to only preselected releases, then it is no longer a transparent Go module source; it is a narrower release-artifact distribution mechanism, and we still need to define what happens for pseudo-versions, unreleased commits, and already-published versions.

I had been assuming, that, since we would not support this on the `source_archive` approach, we would also not support it here: we would just not serve pseudo-versions, unreleased commits or already-published versions. As far as I can understand the Go documentation and golang/go code, this would be okay. It would mean a more limited support compared to "Include generated object files in commit history" but not compared with the `source_archive` approach.

> [The artifact-production system behind the static filesystem] brings back the security concern instead of removing it. A static host for generated module zips is still a high-value module-origin target if it is the thing users trust for `go.opentelemetry.io/...` module contents. The generation and publication pipeline becomes the place where malicious code or mismatched generated artifacts could be introduced, and the static host becomes the place where those artifacts are distributed to downstream users.

On the `source_archive` solution  we also need to have a generation and publication pipeline, namely Github Actions and Github artifacts, which can still be a target of malicious code. If Github Actions and Github Artifacts are deemed as an appropriate solution in terms of security then we could explore making the proxy redirect to zip or plaintext files served on Github. The Go proxy does support redirects so this may be worth exploring (there may be other limitations I am not aware of and I have not tried doing this). Do you foresee any problems with such an approach?

> That is why I do not think it is appropriate for the Collector discussion to turn into "the eBPF projects should solve this by hosting a Go module proxy" unless someone is prepared to own that design and operational burden. 

For me to give an answer to the request for `source_archive`, I need to understand (1) the requirements and problem space and (2) the alternative solutions that were considered. This is why I am taking the time to go through your proposal and understand the constraints. I appreciate your time and patience on explaining the alternatives here, I am not suggesting a particular solution, I am trying to understand your proposal and why it is better than other alternatives.

> Collector maintainers and approvers absolutely get to decide what belongs in OCB. But deciding that OBI, `go.opentelemetry.io/auto`, or other eBPF projects should publish through a new VCS/module-proxy path is a decision for those projects and the relevant OTel infrastructure owners, because they are the ones who would inherit the release, security, storage, and incident-response responsibilities.

The Collector SIG is responsible for the release, security and maintenance of OCB so we also need to fulfill due diligence before accepting a new, bespoke path of component distribution. In short, somebody has to end up assuming some maintenance burden here, so I don't think that should be the focus of the conversation: we should focus on what's best for our user's experience and security.

> A Go module proxy has a different trust model. It changes what bytes are served behind a Go module path/version. That can be fine, but it makes the proxy and its generation pipeline part of the module's publication infrastructure. If the proxy, generated artifact store, or publication pipeline is compromised, misconfigured, or loses historical artifacts, every direct or indirect user resolving through that module path is affected. The Go module checksum database exists specifically because untrusted proxies should not be able to serve different code unnoticed.

The security model of OCB component distribution is to rely on the safeguards from the `go` command, which, for most users, would be to trust on https://proxy.golang.org/ and its checksum database. Users of OCB today already have to trust `go.opentelemetry.io` for providing access to core modules. How would the proxy approach undermine those security guarantees?

I understand we could potentially do things in a better or different way, but I think that is a separate discussion. I want to ground this discussion on comparing the current safeguards when I request, say, `go.opentelemetry.io/collector/exporter/otlpexporter` vs the new modules.
 
> how do we prevent a future release or pseudo-version from being observed by the public proxy before the generated-artifact zip is published

I do not understand this. The public proxy can only observe what is served by the OpenTelemetry proxy in this scenario, so I don't think there is anything to prevent here. Am I misunderstanding something?

> That may be a viable serving layer, but it is not a complete solution. If you believe this is the better path, I think the next useful step is for you to propose it concretely: a design issue or PR for the relevant OTel infrastructure that covers the generation policy, serving layout, release process, pseudo-version behavior, checksum behavior, artifact attestations, SBOM/checksum publication, immutability/retention guarantees, ownership, and failure modes. Ideally it should include a working prototype for at least OBI or `go.opentelemetry.io/auto`, with default Go checksum behavior documented, or an explicit argument for any required `GONOSUMDB`/`GOSUMDB` policy.

I have attempted to sketch out an alternative here. I do not think it is worth continuing with a prototype until we agree on the general idea being a potential path here.

> Until that exists, I do not think [open-telemetry/opentelemetry-collector#15430](https://github.com/open-telemetry/opentelemetry-collector/issues/15430) should be blocked on the possibility of a Go module proxy, 

I agree in principle. I do think however it should be blocked only on making sure everybody understands the problem space and the alternatives considered, which includes resolving this discussion. I hope my answers here can help illuminate the parts that I am failing to understand and the alternative that I want to understand the downsides of.

**Comment by MrAlias** [comment]
Thanks, I think this gets to the core distinction. The Collector SIG should of course decide whether OCB should support `source_archive`, but I do not think a Go module proxy design should become a blocking prerequisite for that decision, period. `source_archive` is a proposal for OCB to consume an explicitly declared, checksum-pinned release artifact. A Go module proxy is a separate publication and operation model for generated module contents. Treating the second design as required before the first can proceed shifts the burden onto the OBI and Go auto-instrumentation SIGs to operate infrastructure they have already decided not to operate.

I also want to be clear about the project-governance side of this. The OBI SIG and the Go auto-instrumentation SIG are the groups with authority over how OBI and `go.opentelemetry.io/auto` publish their generated artifacts. Those SIGs have already discussed this problem space and made a direction-setting decision: publish generated source artifacts as release artifacts rather than make these projects host a Go module proxy/VCS-style publication path. So I do not think this should be framed as "we need to reach a decision" about how the eBPF projects publish their artifacts. A decision has already been made by the SIGs that own that decision. If you do not want to come in and tell those SIGs what they should do, then the next step is to accept that decision as the input to the Collector proposal rather than continuing to relitigate it here.

To make the existing context easier to follow, a lot of the points being re-opened here were already covered earlier in this issue:

- The [issue description](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233) already lists a Go module proxy as an option and calls out that it would require OTel to manage a proxy service, complicate the vanity URL service, add release-process work, and create security attack vectors.
- The module-proxy option was originally raised as requiring a managed OTel service plus a release upload step in [this comment](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2448653025).
- The question of non-official releases/pseudo-versions was raised in [this comment](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2451805042).
- The fallback problem was called out in [this response](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2452124547): fallback to the standard Go proxy or VCS may let users download directly from git history, but those downloads do not have the desired object files.
- Git LFS requiring downstream setup was called out [here](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2453013651) and confirmed from Beyla experience [here](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2682401039).
- Dependency-generation approaches and their `vendor`/`go/pkg/mod` tradeoffs were described [here](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2698298287), [here](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-2887352734), and [here](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-3079576308).
- The bpfgenproxy PoC documented the module-proxy security, `go.sum`, and retention concerns [here](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-3119891483). That PoC is already the dynamic module-proxy shape being sketched again now: a vanity import path using `vcs=mod`, a Go module proxy serving generated module contents, generated `.info`/`.mod`/`.zip` responses for a version, and fallback behavior for anything outside the generated set.

So I do not think the proxy path should be treated as an unexplored alternative that needs to block the Collector proposal. It is a previously considered path with known unresolved operational and security requirements. The latest questions are largely reconstructing the same design space that was already explained in that PoC discussion; the open question was never "can a proxy serve generated bytes?" The open question was who owns that service, how it is secured, how it behaves under normal Go tooling, how it handles checksum canonicalization, and how it reliably serves the generated contents for as long as users can request them.

The release-only constraint is a useful clarification. If the proposed proxy would not serve pseudo-versions, unreleased commits, or already-published versions, then I agree it starts to look closer to the support envelope of `source_archive`. But that also means the proxy is not really a transparent replacement for normal Go module resolution. It is a release-artifact distribution mechanism exposed through the GOPROXY protocol.

If the goal is compatibility with the full behavior of normal Go module tooling, this is not a static-site problem. Go users can request semver releases, pseudo-versions derived from commits, `@latest`, versions already recorded in `go.sum`, versions selected through upgrades or downgrades, and versions reached through transitive dependency resolution. Since a Go module can be downloaded for commits through pseudo-versions, serving generated contents for the full request space means generating those contents either on every relevant upstream push or lazily when a user first requests that version. Either way, there is a generator behind the server, and that generator becomes security-sensitive infrastructure.

That limited scope is a non-starter if it breaks users' standard Go tooling. If the vanity URL points users at a module proxy and that proxy intentionally does not serve parts of the version space users can request, those requests fail unless there is a carefully designed fallback. Breaking ordinary `go get`, `go mod tidy`, `go mod download`, upgrades, downgrades, or transitive dependency resolution for users interacting with `go.opentelemetry.io/...` modules is not a viable option for OpenTelemetry. That is a non-starter, not merely a trade-off. And if fallback goes to VCS or the public proxy for the same module path, we are back to the split-content problem where some versions resolve to generated artifacts and some resolve to repository snapshots without them.

Redirecting to GitHub release assets may help with storage and provenance, but it does not remove the hard parts. The [Go proxy protocol](https://go.dev/ref/mod#goproxy-protocol) still needs module-shaped `.info`, `.mod`, `.zip`, `/@v/list`, and `/@latest` responses, and the `$version.zip` endpoint still needs to return a valid Go module zip with the expected module root and deterministic module hash semantics. Generating a GOPROXY-compatible `.zip` release artifact is probably not the hardest part. The hard parts are accepting the trust position created by the proxy, making the mapping from `module@version` to generated bytes immutable and auditable, ensuring default Go checksum behavior accepts those bytes as canonical, and reliably serving those bytes for as long as Go users can request that version.

Redirects, also, do not remove the need to provide trust in the proxy. The proxy still controls the mapping from `module@version` to the GitHub release assets it redirects to. If that mapping is compromised or misconfigured, users get the wrong module contents even if the bytes ultimately come from GitHub. So a redirect-based design still needs to explain how the mapping is generated, reviewed, published, immutable, monitored, and recoverable.

On the security comparison with GitHub Actions and GitHub release artifacts: I agree both approaches require a generation and publication pipeline. The difference is how consumers bind to the generated bytes. With `source_archive`, the OCB manifest names an explicit URL and pins its SHA-256 digest. In the current OBI release process, those release artifacts are produced by the release workflow, have `SHA256SUMS`, SBOMs, GitHub release asset digests, and GitHub release attestations. If a release artifact changes or a different artifact is substituted, the digest pin in the OCB manifest fails.

With a Go module proxy, the generated artifact becomes the canonical module content for a module path and version. That can be safe only if the [Go checksum path](https://go.dev/ref/mod#authenticating) works with the generated artifact as the canonical content under normal user settings. This is the part the proxy proposal still needs to demonstrate. The bpfgenproxy PoC used `GONOSUMDB` because the generated bytes/checksum behavior was not stable under the normal checksum path. That was not a deployment answer; it was a useful illustration of the severity of getting generation wrong. Once Go records a hash for `module@version` in `go.sum` or the public checksum database records one canonical value, later serving different bytes for the same module version is not a normal recoverable update. Users will keep rejecting the new bytes because the first observed content became the source of truth. Requiring users to disable checksum database verification for `go.opentelemetry.io` would be a failure of the proxy approach, not a viable deployment model.

So, to answer "How would the proxy approach undermine those security guarantees?": it does not necessarily undermine them if the generated module zip is accepted by the public checksum database as the one canonical representation of `module@version` under default Go behavior. But that is the key condition. Current OCB `gomod` dependencies rely on the normal Go module invariant that a module path/version maps to one immutable module hash. A generated-artifact proxy changes where that canonical module content comes from. The proxy proposal needs to prove default Go tooling accepts that cleanly without requiring checksum bypasses.

Before getting into the lower-level technical shape of a release-only proxy, though, the proxy proposal first needs an ownership and security model. The OBI SIG and Go auto-instrumentation SIG have already chosen not to operate a Go module proxy for these artifacts. Without an owner who is willing to operate it, maintain it, monitor it, respond to incidents, retain artifacts, and take a clear security posture for serving generated module contents, the rest of the design is moot. If no group is willing to own those responsibilities, this is not a viable alternative to `source_archive`.

If the module-proxy investigation continues in this issue, it needs to move from questions and sketches into an actual proposal and operation plan. At minimum, that proposal needs to answer:

- Who owns, maintains, monitors, and responds to incidents for the proxy?
- What security posture, availability guarantees, retention guarantees, and release-authority model would the proxy operators commit to?
- Who owns the vanity URL change, proxy behavior, release workflow changes, storage, and retention?
- What exact module paths would be served this way?
- Are they new proxy-only paths, or existing paths with prior VCS/proxy history?
- Are only semver release tags supported?
- If full Go module compatibility is claimed, when are generated module zips created: on every upstream commit, lazily on first request, or only during release?
- What do `/@v/list` and `/@latest` return?
- What happens for pseudo-versions and unsupported versions without breaking standard Go tooling?
- If unsupported versions fall back somewhere else, how do we avoid resolving some versions from generated artifacts and others from VCS/public-proxy snapshots without generated artifacts?
- What exact release assets back `.info`, `.mod`, and `.zip`?
- Is the `.zip` a valid Go module zip generated by the release workflow?
- Does the default Go checksum database accept the generated zip without `GONOSUMDB`, `GOPRIVATE`, `GOSUMDB=off`, or any other checksum bypass?
- What happens if the first generated zip for a `module@version` is wrong after `go.sum` or the public checksum database has recorded it as canonical?
- How are GitHub release attestations, SBOMs, and `SHA256SUMS` tied to the module zip?

Until there is an owned proposal like that, open-telemetry/opentelemetry-collector#15430 should be evaluated on its own merits. The proposal there is not claiming to solve general Go module distribution for generated artifacts. It is saying OCB should be able to consume an explicitly declared, checksum-pinned source archive. If Collector maintainers think that source input is out of scope for OCB, that is a valid decision to make directly. But I do not think that decision should be blocked on a release-only Go proxy idea that does not yet have an operator, a security posture, default checksum-db proof, immutable serving guarantees for generated module zips, or agreement from the SIGs that own the affected eBPF projects to operate or adopt it.

Put differently: if the Collector SIG wants to reject `source_archive` because it does not want OCB to support this class of source input, that is a Collector decision. But it should not be rejected on the basis that the eBPF SIGs have not reached a publication decision, or that they should instead operate a Go module proxy. They have reached a decision, and `source_archive` is a proposal for allowing OCB to consume the release artifacts produced by that decision.

**Comment by mx-psi** [comment]
I have tried to answer your questions below. I think we are slowly getting to the technical aspects I wanted to understand, thanks for your patience on reconstructing the discussion about this!

I think at this point I have the information I needed at a high-level to go back and discuss this with the Collector SIG so we can provide feedback that doesn't just come from me, but if you have the time and patience to bear with me, feel free to clarify any of my replies below as you wish :)

----

> A Go module proxy is a separate publication and operation model for generated module contents. 

Distributing a Go module by the mechanisms supported by the `go` binary which include VCS-based approaches as well as Go module proxy approaches is the only supported OCB component distribution mechanism today. This is what we expect component developers to use and it is the interface that our users expect when including components. We can of course discuss expanding that, but it should be clear that this is not just one mechanism, it is _the_ only supported mechanism today.

> Treating the second design as required before the first can proceed shifts the burden onto the OBI and Go auto-instrumentation SIGs to operate infrastructure they have already decided not to operate.

This decision affects the Collector SIG, therefore it is within our purview to do due diligence before assuming any maintenance burden.

> I also want to be clear about the project-governance side of this. The OBI SIG and the Go auto-instrumentation SIG are the groups with authority over how OBI and `go.opentelemetry.io/auto` publish their generated artifacts. Those SIGs have already discussed this problem space and made a direction-setting decision: publish generated source artifacts as release artifacts rather than make these projects host a Go module proxy/VCS-style publication path. So I do not think this should be framed as "we need to reach a decision" about how the eBPF projects publish their artifacts. A decision has already been made by the SIGs that own that decision. If you do not want to come in and tell those SIGs what they should do, then the next step is to accept that decision as the input to the Collector proposal rather than continuing to relitigate it here.

I think it is fair for me to try to understand the previous discussions given the ask for the Collector SIG to do some work and assume some maintenance burden. I appreciate your patience on summarizing the previous discussion and letting me know about important points in the problem space. If you do not with to continue discussing this that's okay, I can make the decision with the context I have.

> So I do not think the proxy path should be treated as an unexplored alternative that needs to block the Collector proposal.

I agree this is in general terms not an unexplored path. I apologize if any of my comments make it seem like it is, that is not my intention. My intention is to understand the problem space and the constraints, and there were some aspects that were confusing to me and which you have so far kindly provided me pointers to understand.

> That limited scope is a non-starter if it breaks users' standard Go tooling. 

I tend to disagree, but I acknowledge there's room for discussion there. I am however talking here not about general Go tooling but about how to distribute components in a way compatible with OCB, where so far I have not been able to find evidence of breakage (I'd be interested in seeing that if you have examples).

> So, to answer "How would the proxy approach undermine those security guarantees?": it does not necessarily undermine them if the generated module zip is accepted by the public checksum database as the one canonical representation of `module@version` under default Go behavior. But that is the key condition. Current OCB `gomod` dependencies rely on the normal Go module invariant that a module path/version maps to one immutable module hash. A generated-artifact proxy changes where that canonical module content comes from. The proxy proposal needs to prove default Go tooling accepts that cleanly without requiring checksum bypasses.

Thanks, this is helpful and I think this is a key point that needs to be clarified to understand the viability of a Go proxy.

> If no group is willing to own those responsibilities, this is not a viable alternative to `source_archive`.

The counterpart is also true: if the Collector SIG is not willing to accept a non-standard way of consuming Go modules, then we don't have a way to distribute modules, not because of inherent technical limitations, but because no agreement was reached. That's why I think it is important for both parties to understand the problem space.
 
> Put differently: if the Collector SIG wants to reject `source_archive` because it does not want OCB to support this class of source input, that is a Collector decision. But it should not be rejected on the basis that the eBPF SIGs have not reached a publication decision, or that they should instead operate a Go module proxy. They have reached a decision, and `source_archive` is a proposal for allowing OCB to consume the release artifacts produced by that decision.

I would hope we can collaborate together to find a joint solution here and not make decisions separately :)

**Comment by mx-psi** [comment]
Here is a PoC for the option I was proposing: https://go-vanity-proxy.pages.dev/embedmod. This is a Go module served via a fully static Go vanity URL service and Go proxy that just redirects to Github Releases artifacts. As we discussed above this does not serve pseudoversions (the same thing happens with `source_archive`), but otherwise I wasn't able to break it in my tests.

I'd be interested in your thoughts about this and in general, I'd be curious about you trying to test and break this. 

As the page states, the source code is available here https://github.com/mx-psi/static-proxy-poc

**Comment by MrAlias** [comment]
@mx-psi this does not address what I asked for in my previous comment. Have you read what I wrote? I was not asking for another PoC demonstrating that generated module contents can be exposed through the `GOPROXY` protocol. We already had [bpfgenproxy](https://github.com/MrAlias/bpfgenproxy), and I explicitly acknowledged that a Go module proxy is technically possible. Your PoC changes the serving implementation from a proxy process to static Cloudflare redirects backed by GitHub release assets. That demonstrates another implementation of the serving layer, but the serving mechanics were not the part I asked you to make concrete.

In fact, the PoC reinforces the distinction I have been making. CI still generates module contents that are not present in the repository, publishes a canonical module zip to a release, and updates routing metadata that determines which bytes are returned for a module version. Cloudflare may execute only static redirect rules, but that does not remove the trust or operational position of the proxy. It distributes that position across the generation workflow, GitHub release assets, the redirect configuration, and the accounts and credentials that control them.

I therefore do not think asking me to test and break this PoC is the next step. A successful `go get` against a new module path proves the narrow happy path. It does not establish that OpenTelemetry has an owner willing to deploy and operate this design, nor does it establish the security and reliability properties that owner would commit to. Those were the substantive questions in [my previous comment](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1233#issuecomment-4791696687).

The PoC also does not make the pseudo-version comparison with `source_archive` equivalent. `source_archive` is an explicit OCB build input selected by URL and pinned by digest; it does not present itself as the standard Go module source for an existing module path. A `GOPROXY` endpoint behind an OpenTelemetry module path participates in normal Go module resolution. If it intentionally does not serve versions that standard Go tooling can request, the proposal still needs to define behavior for those requests without breaking users or falling back to different module contents. Demonstrating release-only behavior on a brand-new module path with no existing VCS, public proxy, or checksum database history does not answer the cutover and compatibility questions for the existing OBI and Go auto-instrumentation module paths.

What I asked for was an actual OpenTelemetry operation and ownership proposal. Before this can be considered a viable alternative, its proponents need to provide concrete answers to at least the following:

- Which OpenTelemetry SIG or infrastructure team agrees to own the service and the module publication contract?
- Who will implement, review, maintain, monitor, and provide incident response for the release workflow, redirect configuration, vanity URL changes, and artifact storage?
- What security posture will the operators adopt for generation, release authority, credentials, provenance, attestations, checksum database behavior, and compromise recovery?
- What immutability and retention guarantees ensure every published module zip remains available with identical bytes for as long as users can request it, including recovery from accidental deletion or account compromise?
- What availability expectations, monitoring, alerting, backup, disaster recovery, and dependency-failure behavior apply to GitHub Releases, the redirect host, and the vanity URL service?
- How will this be introduced for existing module paths without checksum conflicts, publication races, split contents between the proxy and VCS, or requiring users to change normal Go checksum settings?
- What is the supported-version policy, and what exactly happens when normal Go tooling requests a pseudo-version or another version the release-only proxy does not contain?
- Who owns the design if the current implementers leave, and what is the migration or decommissioning plan once users and `go.sum` files depend on this publication path?

The OBI SIG and Go auto-instrumentation SIG, which own the publication decisions for these projects, have already considered the proxy approach and decided not to take on those responsibilities. A personal PoC does not change that ownership decision. The concrete question is therefore: who is volunteering to own and operate this for OpenTelemetry? If you or the Collector SIG want this design considered as the alternative, the next deliverable needs to be an owned proposal to the relevant OpenTelemetry infrastructure and governance groups, including named operators and commitments for the responsibilities above. The work to turn this into an acceptable OpenTelemetry service belongs with the people proposing that service; it should not be handed back to the eBPF SIGs as another prototype for them to test.

Until such an owner and operation plan exist, this is still an unowned technical prototype, not an organizationally viable alternative to `source_archive`. It should not block [open-telemetry/opentelemetry-collector#15430](https://github.com/open-telemetry/opentelemetry-collector/issues/15430), which should be evaluated directly on whether OCB should support an explicitly declared, checksum-pinned source archive.
